### ISSUE TITLE
Possible false positive "Missing return statement" if return type is Optional[int] etc.

### ISSUE THREAD

**Issue by AndrewUshakov** [python/mypy#9392]
## 🐛 Bug Report

mypy generates error  "Missing return statement" for both functions

```Python
def out() -> Optional[bool]:
    print('ops')


def sqr(x: Optional[int] = None) -> Optional[int]:
    if x is not None:
        return x**2
```

## To Reproduce

1. Create Python file test.py with the content above.
1. Run mypy test.py

## Expected Behavior

in both cases errors can be eliminated by adding just a ```return``` or ```return None``` at the end of both functions, but it is not really needed (does not change program behavior), because by default function without return statement at its end returns None which is a valid return value for Optional[<something>]?

I know about the flag --no-warn-no-return but I do not want to mask other possible more dangerous events.

## Actual Behavior

mypy generates error "Missing return statement" for both functions.

## Your Environment

- Mypy version used: 0.782
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` - file is abcent.
- Python version used: 3.8.5 64-bit
- Operating system and version: Windows 10

**Comment by hauntsaninja** [comment]
Duplicate of #3974, #9000, #9312 and others

**Comment by AndrewUshakov** [comment]
Thank you, I did not find #3974.

1) I agree, that return None should exist in my second function according to PEP8:

"Be consistent in return statements. Either all return statements in a function should return an expression, or none of them should. If any return statement returns an expression, any return statements where no value is returned should explicitly state this as return None, and an explicit return statement should be present at the end of the function (if reachable):"

But mypy error disappears if add just a return at the end of my second function, which is insufficient according to PEP8.  Of course, it is a different small issue (false negative).

2) My first function is valid according to PEP 8 (reason why #3974 was closed was PEP 8 violation).
